api/src/exam-environment/utils/exam.test.ts

```diff
@@ -0,0 +1,307 @@
+//       it is more useful to run the tests against a new generation each time.
+//       This helps ensure the config/logic is _reasonably_ likely to be able to
+//       generate a valid exam.
+//       Another option is to call `generateExam` hundreds of times in a loop test :shrug:
+describe('Exam Environment', () => {
+  describe('checkAttemptAgainstGeneratedExam()', () => {
+    it('should return true if all questions are answered', () => {
+      expect(
+        checkAttemptAgainstGeneratedExam(
+          examAttempt.questionSets,
+          generatedExam
+        )
+      ).toBe(true);
+    });
+
+    it('should return false if one or more questions are not answered', () => {
+      const badExamAttempt = structuredClone(examAttempt);
+
+      badExamAttempt.questionSets[0]!.questions[0]!.answers = [];
+      expect(
+        checkAttemptAgainstGeneratedExam(
+          badExamAttempt.questionSets,
+          generatedExam
+        )
+      ).toBe(false);
+
+      badExamAttempt.questionSets[0]!.questions[0]!.answers = ['bad-answer'];
+      expect(
+        checkAttemptAgainstGeneratedExam(
+          badExamAttempt.questionSets,
+          generatedExam
+        )
+      ).toBe(false);
+
+      badExamAttempt.questionSets[0]!.questions = [];
+      expect(
+        checkAttemptAgainstGeneratedExam(
+          badExamAttempt.questionSets,
+          generatedExam
+        )
+      ).toBe(false);
+    });
+  });
+  xdescribe('checkPrequisites()', () => {
+    // TODO: Awaiting implementation
+  });
+  describe('constructUserExam()', () => {
+    it('should not provide the answers', () => {
+      const userExam = constructUserExam(generatedExam, exam);
+      expect(userExam).not.toHaveProperty('answers.isCorrect');
+    });
+  });
+
+  describe('generateExam()', () => {
+    it('should generate a randomized exam without throwing', () => {
+      const _randomizedExam = timedoutGenerateExam(exam);
+    });
+
+    it('should EVENTUALLY generate an exam matching with the correct number of question sets', () => {
+      const generatedExam = timedoutGenerateExam(exam);
+
+      // { [type]: numberOfType }
+      // E.g. { MultipleChoice: 2, Dialogue: 1 }
+      const generatedNumberOfSets = generatedExam.questionSets.reduce(
+        (acc, curr) => {
+          const eqs = exam.questionSets.find(eqs => eqs.id === curr.id);
+
+          if (!eqs) {
+            throw new Error('Generated question set not found in exam config');
+          }
+
+          return {
+            ...acc,
+            [eqs.type]: (acc[eqs.type] || 0) + 1
+          };
+        },
+        {} as Record<string, number>
+      );
+
+      const configNumberOfSets = exam.config.questionSets.reduce(
+        (acc, curr) => {
+          return {
+            ...acc,
+            [curr.type]: (acc[curr.type] || 0) + curr.numberOfSet
+          };
+        },
+        {} as Record<string, number>
+      );
```

Comment:
As a rule of thumb, if a test includes `reduce` it's too complicated. Tests need to be simple to be useful, so I shouldn't have to step through the logic of a reduce before I can understand what a test is doing.

I get that this is counting the number of question sets of each type, but it's not obvious.

Reply:
What is your suggestion? I cannot hard-code anything here.

Reply:
Okay, so I think I'm getting somewhere. Why discard the `type` information when generating an exam? If you keep it, then the test can look like this:

```ts
const generatedExam = generateExam(exam);
const multipleChoiceSets = generatedExam.questionSets.filter(
  ({ type }) => type === "MultipleChoice"
);
const dialogueSets = generatedExam.questionSets.filter(
  ({ type }) => type === "Dialogue"
);

expect(multipleChoiceSets).toHaveLength(2);
expect(dialogueSets).toHaveLength(1);
expect(generatedExam.questionSets).toHaveLength(3);
```

Reply:
The point of the generated exam is to just keep what is needed to reconstruct an exam. I do not think adding stuff for the sake of a test is a good idea.

Reply:
Okay, so keeping the generated exam as is, we can tidy up as follows. In part because the config is part of the fixture so we don't need to analyse it to find out what question sets it should generate.
```suggestion
      const multipleChoiceIds = exam.questionSets
        .filter(qs => qs.type === 'MultipleChoice')
        .map(qs => qs.id);
      const dialogueIds = exam.questionSets
        .filter(qs => qs.type === 'Dialogue')
        .map(qs => qs.id);

      const generatedExam = generateExam(exam);
      const generatedMultipleChoiceSets = generatedExam.questionSets.filter(
        qs => multipleChoiceIds.includes(qs.id)
      );
      const generatedDialogueSets = generatedExam.questionSets.filter(qs =>
        dialogueIds.includes(qs.id)
      );

      expect(generatedMultipleChoiceSets).toHaveLength(2);
      expect(generatedDialogueSets).toHaveLength(1);
```
Might be better if the ids are exported as fixtures, but this is okay.

Reply:
But now change the exam config, and this unnecessarily breaks.

Reply:
If you want, you can export the "meta" information from the fixtures so that doesn't happen. I'd only bother with that if you change the config regularly, though.